x/incentive/keeper/rewards_savings.go

```diff
@@ -38,30 +38,54 @@ func (k Keeper) AccumulateSavingsRewards(ctx sdk.Context, rewardPeriod types.Mul
 	}
 }
 
-func (k Keeper) SynchronizeSavingsClaim(ctx sdk.Context, owner sdk.AccAddress) {
-	deposit, found := k.savingsKeeper.GetDeposit(ctx, owner)
+// InitializeSavingsReward initializes a savings claim by creating the claim and
+// setting the reward factor indexes
+func (k Keeper) InitializeSavingsReward(ctx sdk.Context, deposit savingstypes.Deposit) {
+	claim, found := k.GetSavingsClaim(ctx, deposit.Depositor)
 	if !found {
-		return
+		claim = types.NewSavingsClaim(deposit.Depositor, sdk.Coins{}, nil)
 	}
-	k.SynchronizeSavingsReward(ctx, deposit)
+
+	rewardIndexes := claim.RewardIndexes
+	for _, coin := range deposit.Amount {
+		globalRewardIndexes, found := k.GetSavingsRewardIndexes(ctx, coin.Denom)
+		if !found {
+			globalRewardIndexes = types.RewardIndexes{}
+		}
+		rewardIndexes = rewardIndexes.With(coin.Denom, globalRewardIndexes)
+	}
+	claim.RewardIndexes = rewardIndexes
+
+	k.SetSavingsClaim(ctx, claim)
 }
 
 // SynchronizeSavingsReward updates the claim object by adding any accumulated rewards
 // and updating the reward index value
-func (k Keeper) SynchronizeSavingsReward(ctx sdk.Context, deposit savingstypes.Deposit) {
+func (k Keeper) SynchronizeSavingsReward(ctx sdk.Context, deposit savingstypes.Deposit, incomingDenoms []string) {
 	claim, found := k.GetSavingsClaim(ctx, deposit.Depositor)
 	if !found {
 		return
 	}
 
-	// Source shares for savings deposits is the deposit amount
-	for _, coin := range deposit.Amount {
-		claim = k.synchronizeSingleSavingsReward(ctx, claim, coin.Denom, coin.Amount.ToDec())
+	// Set the reward factor on claim to the global reward factor for each incoming denom
+	for _, denom := range incomingDenoms {
```

Comment:
Without the `setDiff` fix, this is not safe, because an existing `ukava` deposit will have its reward index set to the global reward index before `synchronizeSingleSavingsReward` is called.